### PR TITLE
🏗Forbid importing sinon

### PR DESCRIPTION
**Pull request by jridgewell**
It's already imported in test suites, and importing your own copy breaks
our special "ensure everything is cleaned up" phase.

Closes #14167.

### CODE REVIEW

**Comment by choumx** [comment]
Strange, I remember having weird test failures when *not* importing `sinon`.

**Comment by rsimha** [comment]
@jridgewell Several tests do specialized setup and teardown stuff via custom sandboxes. Will all that not break? If not, this is a good thing to do.

I also see a sinon error in the travis job for this PR. Maybe rebase on master and rerun tests?

@choumx IIRC, the failures you are referring to happen when a test does creates a custom sandbox and neglects to restore it.

**Comment by rsimha** [comment]
@jridgewell There's a new version of `sinon` that we should probably upgrade to, and it uses a slightly different way to initialize sandboxes. Can you look at https://github.com/ampproject/amphtml/pull/14167#issuecomment-384332942 and maybe incorporate the version upgrade into this PR?

**Comment by jridgewell** [comment]
Yup.

File: package.json [hunk 1]
```diff
@@ -148,7 +148,7 @@
     "request": "2.84.0",
     "rimraf": "2.6.2",
     "rocambole": "0.7.0",
-    "sinon": "4.4.6",
+    "sinon": "4.5.0",
```
Comment (rsimha) [inline package.json]: The latest version is 5.0.0. The release notes don't document it yet, but it's official, and installs fine when you run:
```
yarn add --dev --exact sinon@5.0.0
```

Comment (jridgewell) [reply package.json]: According to [npm](https://www.npmjs.com/package/sinon?activeTab=versions), latest is `4.5.0` and next is `5.0.1-next.1`. 5.0.0 was unpublished 3 months ago (likely an accidental publish).

Comment (rsimha) [reply package.json]: SGTM. We'll have to remove the deprecated usage in our code when 5.0.x gets published, but that can wait.

File: .eslintrc [hunk 2]
```diff
@@ -43,6 +43,7 @@
     "amphtml-internal/no-export-side-effect": 2,
     "amphtml-internal/no-for-of-statement": 2,
     "amphtml-internal/no-global": 0,
+    "amphtml-internal/no-import": 0,
```
Comment (rsimha) [inline .eslintrc]: I assume you want to flag an error here?
```
"amphtml-internal/no-import": 2,
```

**Comment by jridgewell** [comment]
That's [already done](https://github.com/ampproject/amphtml/pull/14779/files#diff-5a0269b2e709fa3dead48cccd5e86d5e).

File: test/_init_tests.js [hunk 3]
```diff
@@ -291,60 +269,45 @@ function warnForConsoleError() {
...
+  consoleSandbox.stub(console, 'info').callsFake(() => {});
+  consoleSandbox.stub(console, 'log').callsFake(() => {});
+  consoleSandbox.stub(console, 'warn').callsFake(() => {});
 }
 
-// Used to restore info, log, and warn level logging after each test.
-function restoreConsoleInfoLogWarn() {
-  consoleInfoLogWarnSandbox.restore();
+if (!sinon.sandbox) {
+  sinon.sandbox = sinon.createSandbox();
```
Comment (rsimha) [inline test/_init_tests.js]: Move this into a `before()` block and do whatever clean up is necessary (like calling `restore()`?) in `after()`.

File: test/_init_tests.js [hunk 4]
```diff
@@ -254,33 +251,14 @@ it.configure = function() {
...
-  }
-  consoleErrorSandbox = sinon.sandbox.create();
-  const originalConsoleError = console/*OK*/.error;
-  consoleErrorSandbox.stub(console, 'error').callsFake((...messages) => {
-    const errorMessage = messages.join(' ').split('\n', 1)[0]; // First line.
+function warnForConsoleError(runner) {
+  const originalConsoleError = console./*OK*/error;
+  const testName = runner.currentTest.fullTitle();
+
+  function record(var_args) {
```
Comment (rsimha) [inline test/_init_tests.js]: Rename `record` to something like `printWarning`.

File: package.json (hunk 1 above)
Comment (rsimha) [reply package.json]: Version 5.0.1 is now out (#14957). There are several new errors that are popping up.

https://travis-ci.org/ampproject/amphtml/jobs/373092869

**Comment by aghassemi** [comment]
Thanks for the Pull Request. Since this PR has been stale for a while, I am closing it. Please feel free to reopen as needed.

**Comment by rsimha** [comment]
`sinon` v6 is out. #16372

Probably worth reopening this?

**Comment by rsimha** [comment]
The changes in this PR were rolled in to #17197.

**Pull request closed without merging**